Implement vfs ilistdir()

dpgeorge opened the pull request:
This PR implements mp_vfs_ilistdir(), and rewrites mp_vfs_listdir() in terms of it.

It also changes VfsFat.listdir() to VfsFat.ilistdir().

Tests are updated and ilistdir() is added to ports uos module.

TODO:
- check what numbers to use for 2nd entry in returned tuple; currently it uses 0x4000 for dirs, 0x8000 for files
- check if FAT FS has something to put in the 3rd entry in the tuple, the "inode"
- test on esp8266 and pyboard
- add tests for calling next() on an exhausted ilistdir iterator, to ensure it raises StopIteration again and doesn't crash

dpgeorge committed:
extmod/vfs: Implement mp_vfs_ilistdir().

uos.ilistdir() is the core function, returning an iterator that yields
3-tuples.  uos.listdir() is implemented in terms of ilistdir().

dpgeorge committed:
extmod/vfs_fat: Replace listdir() with implementation of ilistdir().

VfsFat no longer has the listdir() method.  Rather, if listdir()
functionality is needed then one should use uos.listdir() which will call
VfsFat.ilistdir().

dpgeorge committed:
tests/extmod/vfs: Update tests to reflect new ilistdir() method.

dpgeorge committed:
ports: Add ilistdir in uos module.

pfalcon commented on extmod/vfs.c:
```diff
@@ -286,7 +286,49 @@ mp_obj_t mp_vfs_getcwd(void) {
+typedef struct _mp_vfs_ilistdir_it_t {
+    mp_obj_base_t base;
+    mp_fun_1_t iternext;
+    bool is_str;
```
Maybe put these at the end to not miss further alignment opportunities.

pfalcon commented on extmod/vfs_fat_misc.c:
```diff
@@ -34,51 +34,64 @@
+            // stop on error or end of dir
+            break;
+        }
+        if (fn[0] == '.' && (fn[1] == 0 || (fn[1] == '.' && fn[2] == 0))) {
```
So, on unix, ilistdir() returns `.` and `..`, and I guess it's correct behavior, especially that now with C implementation of listdir(), it's not a trouble to filter them out there.

dpgeorge replied:
I was going by CPython's os.listdir(), which *doesn't* return . and ...  Neither does os.scandir().  And because . and .. are specific to the VFS (probably VfsSPIFFS , if it existed, wouldn't have such entries).  So I think it's up to the VFS to discard them.

pfalcon replied:
But then we don't have portable way to access current and parent dir entries, can't even make ls -la workalike in MicroPython. The idea is that ilistdir is a superset of listdir. And it should just return whatever directory entries underlying FS has. If they're not there - ok, it's up to a user to deal with that.

pfalcon replied:
On the other hand, in apps which don't do ls -la (99% of apps), filtering out . and .. is a chore... So need to think well about implications of filtering, and then again update unix port.

dpgeorge replied:
>  and then again update unix port.

Unix port won't need updating because all FS functionality in unix/moduos.c will be deleted and the extmod/vfs_posix versions used instead.

pfalcon replied:
Makes sense ;-). OTOH, re: another issue, the current patch still does:

~~~
t->items[1] = MP_OBJ_NEW_SMALL_INT(dirent->d_type);
~~~

pfalcon commented:
> check what numbers to use for 2nd entry in returned tuple; currently it uses 0x4000 for dirs, 0x8000 for files

Well, unix port (on Linux) uses small numbers - 4, 8...

pfalcon commented:
This looks great, definitely let's target this for 1.9.

dpgeorge commented:
> Well, unix port (on Linux) uses small numbers - 4, 8...

These constants need to be known to the end user so they can distinguish files from dirs.  Using the same value as stat() returns seems sensible to me (which is what the current, large numbers are).


> This looks great, definitely let's target this for 1.9.

Yes.

pfalcon commented:
> Using the same value as stat() returns seems sensible to me (which is what the current, large numbers are).

Except that POSIX think different. Values returned by ilistdir() are based on readdir() and struct dirent. Note that POSIX actually mandates only name and inode: http://pubs.opengroup.org/onlinepubs/7908799/xsh/dirent.h.html, type is a common useful extension to avoid stat(), if it's 0, a user still has to do it. Values of type are: https://linux.die.net/man/3/readdir .

We can use stat's values instead, but should think well about that, and then synchronously update unix version.

dpgeorge commented:
I added some update commits.

> check what numbers to use for 2nd entry in returned tuple; currently it uses 0x4000 for dirs, 0x8000 for files

I changed VFS core and VfsFat to use 4 and 8, to be consistent with unix.  Still need to decide on this, whether to change everything to the same values returned by stat.

> check if FAT FS has something to put in the 3rd entry in the tuple, the "inode"

No it doesn't, there's no useful info in the FILINFO struct that can be used.

> test on esp8266 and pyboard

Done, it works as expected.  listdir() still works the same as before.

> add tests for calling next() on an exhausted ilistdir iterator, to ensure it raises StopIteration again and doesn't crash

Done.

Main thing left to do is decide where to filter out "." and "..".

pfalcon commented on extmod/vfs_fat_misc.c:
```diff
@@ -66,10 +66,10 @@ STATIC mp_obj_t mp_vfs_fat_ilistdir_it_iternext(mp_obj_t self_in) {
         }
         if (fno.fattrib & AM_DIR) {
             // dir
-            t->items[1] = MP_OBJ_NEW_SMALL_INT(0x4000); // S_IFDIR
```
But we were in the process of discussing this! I was waiting for your turn in considering of possible problems with not using DT_* types, and if none, then using stat's values makes good sense!

pfalcon commented:
> I changed VFS core and VfsFat to use 4 and 8, to be consistent with unix. Still need to decide on this, whether to change everything to the same values returned by stat.

So, struct dirent's `DT_*` values seem to be fully parallel to stat's `S_*`. Please ponder of any issues with using stat values on your side, and if none, let's stick with stat's.

> No it doesn't, there's no useful info in the FILINFO struct that can be used.

We can fix that later ;-).

> Main thing left to do is decide where to filter out "." and "..".

So again, please think of more good reasons to include them. If none, let's drop them.


Thanks!

dpgeorge commented:
> Please ponder of any issues with using stat values on your side, and if none, let's stick with stat's.

VfsFat needs translation whatever we choose.  Listing mount points in the root dir will just use whatever value we pick.  It's only unix/posix that has some predefined notion/values.  In the end I don't think it matters, so long as we have the option of providing constants (eg os.ILIST_DIR and os.ILIST_FILE) so that users don't need to remember that 0x8000 is a directory, for example.  In this respect it would be easier that the values are the same as for stat(), so the same constants could be used.

> please think of more good reasons to include them. If none, let's drop them.

I can only think of reasons to exclude "." and "..": 1) having them makes the ilistdir() tests more verbose; 2) having them means i-listing the root dir will include them only if a FS is mounted there which has such entries (eg FAT, POSIX), and not include them otherwise, which may lead to special code to account for such a (minor) inconsistency.

"." and ".." are of course very useful entities when working with directories, but I don't think there's ever a case where I care that they are in a directory listing; I know such entries always exist.

The only use for "." and ".." I can think of in ilistdir() is that you can get the inode number.  All the other info is already known (that they exist and that they are directories).  But if you want the inode number you can just use stat().

pfalcon commented:
> In this respect it would be easier that the values are the same as for stat(), so the same constants could be used.

Right, the question we decide whether ilistdir() is modeled on POSIX readdir (well, `DT_*` values aren't even POSIX!) or not. I guess, "not" is pretty good answer, so let's use the same constants as stat().

>  But if you want the inode number you can just use stat().

Right. So let's skip "." and ".." and have listdir() and ilistdir() consistent in this respect.

Thanks!

dpgeorge commented:
Ok, all comments above are taken into account (stat values are used and "." and ".." are not included in ilistdir() output) and this is now merged.

dpgeorge commented:
Basic docs for ilistdir were added in 6d221fe284f977dbf55fd4668efae281b045c237

dpgeorge commented:
It turns out that FatFS already filters out . and ..  So I removed the redundant check in the VFS layer in f95e4e77823919dfe82b6711f1d25d2d8c5008fc

pfalcon commented:
Sounds good, if it's just FatFs' VFS implementation.

dpgeorge commented:
> if it's just FatFs' VFS implementation

Yes, the change was just in VfsFat.
